be/src/exec/spill/spill_components.cpp

```diff
@@ -0,0 +1,352 @@
+            break;
+        }
+        ASSIGN_OR_RETURN(auto chunk, reader->restore(_runtime_state, SyncTaskExecutor{}, guard));
+        restore_rows += chunk->num_rows();
+        if (chunk->is_empty()) {
+            continue;
+        }
+        auto hash_column = down_cast<SpillHashColumn*>(chunk->columns().back().get());
+        const auto& hash_data = hash_column->get_data();
+        // hash data
+        std::vector<uint32_t> shuffle_result;
+        shuffle_result.resize(hash_data.size());
+        size_t left_channel_size = 0;
+        for (size_t i = 0; i < hash_data.size(); ++i) {
+            shuffle_result[i] = hash_data[i] >> current_level & 0x01;
+            left_channel_size += !shuffle_result[i];
+        }
+        size_t left_cursor = 0;
+        size_t right_cursor = left_channel_size;
+        std::vector<uint32_t> selection(hash_data.size());
+        for (size_t i = 0; i < hash_data.size(); ++i) {
+            if (shuffle_result[i] == 0) {
+                selection[left_cursor++] = i;
+            } else {
+                selection[right_cursor++] = i;
+            }
+        }
+
+#ifndef NDEBUG
+        for (size_t i = 0; i < left_cursor; i++) {
+            DCHECK_EQ(hash_data[selection[i]] & left_partition->mask(),
+                      left_partition->partition_id & left_partition->mask());
+        }
+
+        for (size_t i = left_cursor; i < right_cursor; i++) {
+            DCHECK_EQ(hash_data[selection[i]] & right_partition->mask(),
+                      right_partition->partition_id & right_partition->mask());
+        }
+#endif
+
+        if (left_channel_size > 0) {
+            ChunkPtr left_chunk = chunk->clone_empty();
+            left_chunk->append_selective(*chunk, selection.data(), 0, left_channel_size);
+            size_t old_rows = left_partition->num_rows;
+            RETURN_IF_ERROR(_spill_partition(spill_ctx, left_partition, [&](auto& consumer) {
+                consumer(left_chunk);
+                left_partition->num_rows += left_chunk->num_rows();
+                return Status::OK();
+            }));
+            DCHECK_EQ(left_channel_size, left_partition->num_rows - old_rows);
+            // left->spill_writer->t_spill(state, left_chunk, SyncTaskExecutor{}, guard);
+        }
+        if (hash_data.size() != left_channel_size) {
+            ChunkPtr right_chunk = chunk->clone_empty();
+            right_chunk->append_selective(*chunk, selection.data(), left_channel_size,
+                                          hash_data.size() - left_channel_size);
+            size_t old_rows = right_partition->num_rows;
+            RETURN_IF_ERROR(_spill_partition(spill_ctx, right_partition, [&](auto& consumer) {
+                consumer(right_chunk);
+                right_partition->num_rows += right_chunk->num_rows();
+                return Status::OK();
+            }));
+
+            DCHECK_EQ(hash_data.size() - left_channel_size, right_partition->num_rows - old_rows);
```

Comment:
release chunk earily here to avoid memory useage

Reply:
The declaration cycle of chunk is only valid within the while loop and does not need to be destructured in advance